### PR TITLE
Reimplement capture toolbar in Qt rather than ImGui.

### PR DESCRIPTION
**Pull request by pierricgimmig**
The toolbar now auto-resizes based on the window width. Fixes b/158244053. 

Note that I removed the process name.  Not sure it adds tremendous value and it takes a lot of screen real estate. We could always have it in the title bar similar to what we see when we load a capture.

qt_toolbar

### CODE REVIEW

**Comment by dpallotti** [comment]
Just from the screenshot: What about taking it out of the "capture" tab and positioning it directly under the menu bar like toolbars normally are? This way it would always use the full width of the window (I don't think it's a problem if we lose a tiny bit of vertical space on the right side). Otherwise I have a feeling that when the capture tab is not that wide it would take several lines. Also some buttons don't only apply to the capture view, for example load, save, feedback.

**Comment by beckerhe** [comment]
I like Davide's idea of using the designer for creating the toolbar.

This would probably require to move the icon-loading-code to Qt's qrc system for seeing and setting the icons in the designer.

File: OrbitQt/orbitmainwindow.cpp [hunk 1]
```diff
@@ -217,6 +223,107 @@ OrbitMainWindow::OrbitMainWindow(QApplication* a_App,
...
+  toolbar->addAction(ui->actionStop_Capture);
+  toolbar->addAction(ui->actionClear_Capture);
+  toolbar->addAction(ui->actionOpen_Capture);
+  toolbar->addAction(ui->actionSave_Capture);
+  toolbar->addAction(ui->actionHelp);
+  toolbar->addAction(ui->actionFeedback);
+
+  // Filter tracks.
+  toolbar->addWidget(CreateSpacer(toolbar));
+  toolbar->addAction(CreateDummyAction(icon_filter, toolbar));
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitQt/orbitmainwindow.cpp:291]: > Just from the screenshot: What about taking it out of the "capture" tab and positioning it directly under the menu bar like toolbars normally are? This way it would always use the full width of the window (I don't think it's a problem if we lose a tiny bit of vertical space on the right side). Otherwise I have a feeling that when the capture tab is not that wide it would take several lines. 

I really want to minimize vertical space taken up by UI. I initially tried having the toolbar at the top level, and didn't like how it looked but more importantly I didn't think it made sense for all the actions to be available globally.

> Also some buttons don't only apply to the capture view, for example load, save, feedback.

Those buttons all apply to the capture view, did you mean not "only" apply to the capture view? Some of those actions are already present in the main menu, so they are accessible from anywhere. Inversely, having the toolbar at the top level would expose functionality that is not at all pertinent to any other tabs than the capture view (start/stop capture without visual feedback is a risk; filter tracks and the timer label wouldn't make sense either in any other view.

> Can some of this code be replaced with definitions in the .ui file (e.g.: adding the actions and the other widgets to the toolbar)? That's what is done for the menu.

That's how it was initially done but I found that as soon as you need to do something a little bit special (adding a spaces between elements for example) then you'd have to add some code.  Also, in Qt Creator, you can't add a QLineEdit to a toolbar, this needs to be done in code, or possibly by editing the xml manually which I'm not a fan of.  I found that we have way more control in code.

> Isn't it possible to add an icon to the toolbar directly?

You can't add a QIcon directly to a toolbar.  I tried adding a QPushButton instead of the action, but the look is different.

File: OrbitQt/orbitmainwindow.cpp [hunk 2]
```diff
@@ -492,11 +599,35 @@ void OrbitMainWindow::OnTimer() {
 
   // Output window
   this->ui->plainTextEdit->OnTimer();
+
+  if (timer_label_) {
+    timer_label_->setText(GOrbitApp->GetCaptureTime().c_str());
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitQt/orbitmainwindow.cpp]: Done

File: OrbitQt/orbitmainwindow.cpp [hunk 3]
```diff
@@ -217,6 +223,107 @@ OrbitMainWindow::OrbitMainWindow(QApplication* a_App,
...
+
+  // Filter functions.
+  toolbar->addWidget(CreateSpacer(toolbar));
+  toolbar->addAction(CreateDummyAction(icon_search, toolbar));
+  filter_functions_line_edit_ = new QLineEdit();
+  filter_functions_line_edit_->setClearButtonEnabled(true);
+  filter_functions_line_edit_->setPlaceholderText("filter functions");
+  SetFontSize(filter_functions_line_edit_, kFontSize);
+  toolbar->addWidget(filter_functions_line_edit_);
+  connect(filter_functions_line_edit_, SIGNAL(textChanged(const QString&)),
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitQt/orbitmainwindow.cpp]: Done.

File: OrbitQt/orbitmainwindow.cpp [hunk 4]
```diff
@@ -217,6 +223,107 @@ OrbitMainWindow::OrbitMainWindow(QApplication* a_App,
...
+
+QAction* CreateDummyAction(const QIcon& icon, QObject* parent) {
+  QAction* action = new QAction(icon, "", parent);
+  action->setDisabled(true);
+  return action;
+}
+
+QIcon GetIcon(const std::string& icon_name) {
+  return QIcon(Path::JoinPath({Path::GetIconsPath(), icon_name}).c_str());
+}
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitQt/orbitmainwindow.cpp:247]: Done.

File: OrbitGl/CaptureWindow.cpp [hunk 5]
```diff
@@ -1077,29 +896,4 @@ void CaptureWindow::RenderTimeBar() {
...
 //-----------------------------------------------------------------------------
-void CaptureWindow::Initialize() {
-  GlCanvas::Initialize();
-  LoadIcons();
-}
-
-//-----------------------------------------------------------------------------
-uint32_t LoadIcon(const char* name) {
-  std::string icon_path = Path::GetExecutablePath() + "icons/" + name;
-  return LoadTextureFromFile(icon_path.c_str());
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/CaptureWindow.cpp:1088]: I thought about removing it, but I think it could be useful in the future. Actually let me remove it in another PR, we can alway revert the removal commit if we need it again.

**Comment by dpallotti** [comment]
> I really want to minimize vertical space taken up by UI.

Is vertical space more important in the capture view or in the right-hand-side tabs? How does the toolbar scale when it doesn't fit in width in the capture view? If it ends up taking new lines, then more vertical space is taken from the capture view.

> I didn't think it made sense for all the actions to be available globally.

> having the toolbar at the top level would expose functionality that is not at all pertinent to any other tabs than the capture view (start/stop capture without visual feedback is a risk; filter tracks and the timer label wouldn't make sense either in any other view.

Of course. Some actions would have to be hidden based on whether the "capture" tab is currently selected, but some others (like load and save&mdash;honestly for feedback there's a dedicated label a two centimetres above a so that button for me could just go) should always be available. 

Also start and stop capture would still provide a feedback as the buttons would at least change state and the timer would start/stop. If it was possible to hit the start button while in the "home" tab and have that switch automatically to the "capture" tab, that would even save users from having to always click on the "capture" tab first.

> Some of those actions are already present in the main menu, so they are accessible from anywhere.

Based on this reasoning some actions should just not be on the toolbar at all. 

Bottom line of what I'm saying it that having a toolbar with global actions and actions that even if not global are still main actions of Orbit and might be useful regardless of what tab is selected (capturing) as part of a tab is weird. Normal positioning of a toolbar in basically all desktop applications is right below the menu taking all the width (which also mean that it scales only with the window, not with the window and the left pane).

File: OrbitQt/orbitmainwindow.cpp [hunk 6]
```diff
@@ -217,6 +223,107 @@ OrbitMainWindow::OrbitMainWindow(QApplication* a_App,
...
+
+void OrbitMainWindow::SetupCaptureToolbar() {
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitQt/orbitmainwindow.cpp:250]: > in Qt Creator, you can't add a QLineEdit to a toolbar, this needs to be done in code, or possibly by editing the xml manually which I'm not a fan of

Ah, so far I've always edited the xml manually...

File: OrbitGl/CaptureWindow.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (beckerhe) [reply OrbitGl/CaptureWindow.cpp:1088]: Sounds good!

**Comment by pierricgimmig** [comment]
> How does the toolbar scale when it doesn't fit in width in the capture view?

It doesn't take new lines, toolbar elements disappear "gracefully" unlike the previous implementation.

> Bottom line of what I'm saying is that having a toolbar with global actions and actions that even if not global are still main actions of Orbit and might be useful regardless of what tab is selected (capturing) as part of a tab is weird.

Maybe we should get input from UX here.  I would suggest having the start capture action globally mapped to F5 and automatically focusing on the capture tab when pressed. Stop capture would be mapped to Shift-F5. We'd still have the buttons in the capture view.  Maybe we should have a "Capture" menu as well. I still don't see the value of having a global toolbar.

> Normal positioning of a toolbar in basically all desktop applications is right below the menu taking all the width

I disagree. Firstly because I really think it's a useless waste of space, but also because modern applications such as Visual Studio Code for example are moving away from this pattern by having toolbars that are contextual, take up minimal space, and can definitely be per-tab.

If we really wanted to have a global toolbar, we'd still want to have the capture specific toolbar, something like this:

global_toolbar

I'd suggest moving this conversation to another channel and submitting the change for now as it is an improvement over the previous implementation and is basically a 1:1 mapping in functionality.

**Comment by dpallotti** [comment]
I still have a lot of thoughts about your reply but we should take it offline as it would probably be mostly repeating what I said before and nobody will change his opinion, so I'll focus on the main point.

> It doesn't take new lines, toolbar elements disappear "gracefully" unlike the previous implementation.

> it is an improvement over the previous implementation and is basically a 1:1 mapping in functionality.

Toolbar elements shouldn't just disappear when the width of the Window is much more (usually double) than the space available for the toolbar, and this is the main problem that should be solved, not how graceful they disappear, that's why it shouldn't be a 1:1 mapping.

The suggestion of having a single global toolbar with some controls appearing when the "capture" tab is selected killed two birds with one stone: it allowed more space for the toolbar and provided basic global actions (including the very main action of the application) always one click away, regardless of the current tab.

**Comment by ronaldfw** [comment]
Not touching on points that are in the discussion, but pragmatically I would say that this PR is an improvement over the existing implementation. Even if we can't agree right now on how to do the design of the toolbar (or multiple toolbars), this PR should be merged and we can then iterate on the design.
